src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala

```diff
@@ -11,27 +12,47 @@ trait LogicalTreeNode {
   def getOMComponents(resourceBindingsMap: ResourceBindingsMap, children: Seq[OMComponent] = Nil): Seq[OMComponent]
 }
 
+class RootLogicalTreeNode extends LogicalTreeNode {
+def getOMComponents(resourceBindingsMap: ResourceBindingsMap, children: Seq[OMComponent] = Nil): Seq[OMComponent] = children
+}
+
 object LogicalModuleTree {
+  val resourceScopes = collection.mutable.Map[LogicalTreeNode, () => ResourceBindingsMap]()
+
+  val root = new RootLogicalTreeNode()
+
   private val tree: mutable.Map[LogicalTreeNode, Seq[LogicalTreeNode]] = mutable.Map[LogicalTreeNode, Seq[LogicalTreeNode]]()
-  def add(parent: LogicalTreeNode, child: => LogicalTreeNode): Unit = {
+
+  def addResourceScope(ltn: LogicalTreeNode, rbm: () => ResourceBindingsMap): Unit = {
+    require(! resourceScopes.contains(ltn))
+    resourceScopes.put(ltn, rbm)
+  }
+
+  def add(parent: LogicalTreeNode, child: => LogicalTreeNode, resourceBindingsMap: Option[() => ResourceBindingsMap] = None): Unit = {
     val treeOpt = tree.get(parent)
     val treeNode = treeOpt.map{
```

Comment:
Add a check that the parent does not contain the child already.

Add a check that the child has not been added to the tree more than once.

This can be done in the map add function.

This means when adding a parent child edge in the map that the child should not appear in any other parent child list. This can be checked with a second map of all children.